src/System.Net.Http/tests/FunctionalTests/HttpCookieProtocolTests.cs

```diff
@@ -0,0 +1,413 @@
+                handler.CookieContainer = CreateSingleCookieContainer(url);
+
+                using (HttpClient client = new HttpClient(handler))
+                {
+                    Task<HttpResponseMessage> getResponseTask = client.GetAsync(url);
+                    Task<List<string>> serverTask = LoopbackServer.ReadRequestAndSendResponseAsync(server);
+                    await TestHelper.WhenAllCompletedOrAnyFailed(getResponseTask, serverTask);
+
+                    List<string> requestLines = await serverTask;
+
+                    Assert.Contains($"Cookie: {s_expectedCookieHeaderValue}", requestLines);
+                    Assert.Equal(1, requestLines.Count(s => s.StartsWith("Cookie:")));
+                }
+            });
+        }
+
+        [Fact]
+        public async Task GetAsync_SetCookieContainerMultipleCookies_CookiesSent()
+        {
+            var cookies = new Cookie[]
+            {
+                new Cookie("hello", "world"),
+                new Cookie("foo", "bar"),
+                new Cookie("ABC", "123")
+            };
+
+            await LoopbackServer.CreateServerAsync(async (server, url) =>
+            {
+                HttpClientHandler handler = CreateHttpClientHandler();
+
+                var cookieContainer = new CookieContainer();
+                foreach (Cookie c in cookies)
+                {
+                    cookieContainer.Add(url, c);
+                }
+
+                handler.CookieContainer = cookieContainer;
+
+                using (HttpClient client = new HttpClient(handler))
+                {
+                    Task<HttpResponseMessage> getResponseTask = client.GetAsync(url);
+                    Task<List<string>> serverTask = LoopbackServer.ReadRequestAndSendResponseAsync(server);
+                    await TestHelper.WhenAllCompletedOrAnyFailed(getResponseTask, serverTask);
+
+                    List<string> requestLines = await serverTask;
+
+                    string expectedHeader = "Cookie: " + string.Join("; ", cookies.Select(c => $"{c.Name}={c.Value}").ToArray());
+                    Assert.Contains(expectedHeader, requestLines);
+                    Assert.Equal(1, requestLines.Count(s => s.StartsWith("Cookie:")));
+                }
+            });
+        }
+
+        [Fact]
+        public async Task GetAsync_AddCookieHeader_CookieHeaderSent()
+        {
+            if (IsNetfxHandler)
+            {
+                // Netfx handler does not support custom cookie header
+                return;
+            }
+
+            await LoopbackServer.CreateServerAsync(async (server, url) =>
+            {
+                HttpClientHandler handler = CreateHttpClientHandler();
+                using (HttpClient client = new HttpClient(handler))
+                {
+                    HttpRequestMessage requestMessage = new HttpRequestMessage(HttpMethod.Get, url);
+                    requestMessage.Headers.Add("Cookie", s_customCookieHeaderValue);
```

Comment:
Can you add a variation for multiple values/cookies set?

Reply:
Sure